docs/developer/architecture/interaction-with-k8s-from-salt.rst

```diff
@@ -0,0 +1,96 @@
+Interaction with Kubernetes API from Salt
+=========================================
+
+Context
+-------
+
+In MetalK8s we need to interact with Kubernetes API from Salt in order to
+create and edit various objects. Since the Salt Kubernetes module is not
+flexible enough (at least today) we decided to build our own Salt module that
+will cover all we need in MetalK8s. This Salt module relies on python
+Kubernetes lib so that we have a proper python class for every Kubernetes
+object and we can validate the object content without any call to the actual
+Kubernetes API.
+
+But Python Kubernetes lib is a bit late on Kubernetes versions (at the
+time of writing, the latest python-kubernetes is for k8s-1.18 and k8s 1.22.1
+just came out) so it means that some Kubernetes objects from the latest
+versions cannot be managed by this python lib.
+
+Here, we want to solve this issue and make sure when a new Kubernetes
+version comes out that we can easily support interaction with all the
+Kubernetes API from Salt.
+
+Design Choices
+--------------
+
+To have proper support of the full Kubernetes API we want to use directly
+``kubectl``, which is the CLI build as part of Kubernetes.
+
+The advantage of using ``kubectl`` is that we do not have anything to do in
+order to support interaction with the new Kubernetes API and nothing (or
+almost) to maintain in our Salt module that uses this CLI.
+It also means that we need to use a CLI in order to do some “HTTP query”
+to Kubernetes API, which adds an extra layer, but it’s a small inconvenience
+compare to the benefits.
+
+Rejected Design Choices
+-----------------------
+
+Contribute to python-kubernetes Lib
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We could contribute to python-kubernetes in order to make it compatible with
+the latest versions of Kubernetes.
+
+This approach was rejected because python-kubernetes start to be really late
+on Kubernetes versions, and this work, likely, needs to be done for every new
+Kubernetes version.
+
+Direct HTTP Query to Kubernetes API from Python
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We could build a Salt module that directly interacts with Kubernetes API the
+way we want so that we do not rely on python-kubernetes but only implement the
+interactions we need.
+
+This approach was rejected because it means this Salt module needs to be
+maintained for every new Kubernetes version and for every API/object we want
+to support.
```

Comment:
Why's that? Given a K8s object (any, really) you can construct the URL to PUT/POST to, or GET from. The API is very regular. There should be no need to make any changes to such 'generic' client infrastructure.

It's quite likely the upstream Python K8s client already has everything you need: where it falls short right now is the models/classes representing all kinds of K8s objects (generated from OpenAPI specs). Underlying, there's the functionality to call into the HTTP API which we could re-use, generically, for any kind of object (since we no longer rely on the models/classes to be kept in sync).

Reply:
https://github.com/kubernetes-client/python-base/tree/master/dynamic (H/t @gdemonet)

Reply:
You're right, it would be a better approach, I will check this and update this PR